Added pipeline installation process

Alopalao opened the pull request:
Closes #4 

### Summary

Added listener methods

Added pipeline installation process. There are few additions to the process:
- New `status` to pipelines, `enabling`. This value indicates that the pipeline is in process to be enabled but not finished yet. The status change from `enabling` to `enabled` is done by the code and not the user.

- Added basic pipeline to `settings`, this resembles the pipeline setup before `of_multi_table` installation.

- New 'status' to pipelines, `disabling`. This value indicates that the pipeline is in process to be disabled but not finished yet.

- New added status value `_error` which will be added to the pipeline `status` in case an related to `of_multi_table` has occurred. These can be a miss flows triggered `OFPT_ERROR`.

Other PR are needed to this commit. [mef_eline](https://github.com/kytos-ng/mef_eline/pull/325)(no longer necessary so far), [of_lldp](https://github.com/kytos-ng/of_lldp/pull/80) and [flow_manager](https://github.com/kytos-ng/flow_manager/pull/150)

### Local Tests

Tested this commit with two different processes in mind:

- Kytos console start: When this NApp starts, the pipeline could be:

1. `enabling`: The enabling of a pipeline did not finished in a previous run. The enabling process will start over again. Flows are still installed by consistency check.
2. `enabled`: This means that a pipeline has been enabled. Only events will be sent to NApps with pipeline setup
3. Default pipeline: Send events with default pipeline setup.
4. `disabling`: A pipeline did not finished the disabling process before. So now the default pipeline is going to be applied.

- Pipeline changed: This is triggered by Requests. The pipeline change could be:

1. Default - Enabling - Enabled: This process is started by `"/v1/pipeline/{pipeline_id}/enable" POST`.
    - Update: Enabling a pipeline can be done even if another pipeline is enabled.
2. Enabled - Disabling - Disabled - Default: This process is started by `"/v1/pipeline/{pipeline_id}/disable" POST`.
    - Update: To rectify default pipeline, the user can try to disable (and all pipelines are disabled) any pipeline and the process of activating a default pipeline will resume

NOTE:
In a case a switch cannot not accept a flow with a certain table id (OFPT_ERROR) has not been tested.

Alopalao commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+                      f"code {response.status_code}.")
+            # Problem here: pipeline stays as 'enabling', error needs
+            # to be handled. Events have been sent and NApps are probably
+            # working in a different pipeline setup
```
Error here is not handled. I have two solutions for this:

- I can get the flows before sending any `enable_table` events and stored them in a self variable. Delete the content of the self variable after a pipeline has finished enabling or an error occurs.

- This part of the code is reached when enabling pipeline and disabling pipeline:

    1. Enabling pipeline error is easy to handle, resend `enable_table` with default pipeline and add extra value `error` to `self.required_napps`.

    2. When disabling a pipeline, the NApp needs to return to default. I would need to get the most recent disabled pipeline enable it again and resend `enable_table` with extra value `error` to `self.required_napps`. 

The extra value is added so `get_flows_to_be_installed()` is not called, see lines 123-127.
I came with these solutions while writing this comment so maybe there is a better one.

viniarck replied:
Very well thought @Alopalao. 

> Enabling pipeline error is easy to handle, resend enable_table with default pipeline

This sounds really great to me, it would make this NApp very reliable, since we know for a fact that falling back to the default pipeline would ensure a smooth operation. 

> When disabling a pipeline, the NApp needs to return to default. I would need to get the most recent disabled pipeline enable it again and resend enable_table with extra value error to self.required_napps.

Theoretically, if a pipeline was successfully enabled, when disabling it you cannot get any of the unrecoverable errors from opft_error assuming of course we'll only send coherent payload, so when disabling a pipeline always returning to the default sounds good to me. 

> The extra value is added so get_flows_to_be_installed() is not called, see lines 123-127.
I came with these solutions while writing this comment so maybe there is a better one.

Yes, there is @Alopalao . If instead of using requests you use `kytos.flow_manager.flows.(install|delete)` then it become simpler to avoid to deal with retries or timeouts, also you can leverage the `force=True` option, which will make `flow_manager` to deal with any socket error that can happen and eventually the consistency check will install or remove it. The only error you'd need to handle would be subscribing to `kytos/flow_manager.flow.error`, and since you'd be using `force=True`, then on `kytos/flow_manager.flow.error`, you only need to check if ```error_exception`` key is set, if it is then it's an OFPT_ERROR, so it's something you can't recover from and you'll need to rollback again to the basic pipeline. Also, another benefit of using events to install or delete flows is that it'd be easier to make it production-grade, notice that you can be dealing with 1000s of flows here which currently the responsibility is on the callers to pace it accordingly not to overwhelm the switches, with that said I'd recommend that you stick with a batched/interval approach [like mef_eline does grouping by switches](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L729-L746), so to send for instance 1000 flows it would take 10 secs more or less, which is reasonable since installing a pipeline isn't keep happening frequently, and in fact we can expect to deal with 1000 ish flows in prod in this iteration, and mef_eline has also tested these parameters heavily on convergence tests so we know that switches can handle this sort of load without getting too overwhelemed, so could you also replace requests with the event interface when sending flows and parametrizing the batch_size/interval like mef_eline does, and later on we might further tune these parameters?

Alopalao replied:
Added mef_eline method to group flows by batches.
Also, added the handling of the defined error. It returns successfully to a previous state. Only one thing, if a user creates a napp called `error`, and a listener is added for `enable_table` this solution will not work and it will enter into a loop of trying to get flows from flow manager until a 200 is returned. Do we have a list of not allowed names for NApps?

viniarck replied:
@Alopalao good observation. I wouldn't expect our core team or any other developer to create a NApp called `error`, but to same `of_multi_table` shouldn't be vulnerable to this. This is a sign that we should try to implement a slightly more robust mechanism. Can you come up with another additional variable to keep track when an error happens and always to clean it up whenever needed?

Alopalao commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+                'instructions': table.get('instructions', instruction),
+                'owner': 'of_multi_table',
+                'table_group': 'base',
+                'table_id': table['table_id'],
```
I am installing EPL as miss flow entries when a miss flow is missing in the pipeline. Planning to install a miss flow only when the pipeline has it.

viniarck replied:
Right, yes it should only be installing when `table_miss_flow` isn't an empty. I'll review this part again in the feature when you update it.

Alopalao replied:
Now only installing miss flows if any is defined in the pipeline.

Alopalao commented on main.py:
```diff
@@ -118,9 +313,56 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
+        switch_id = event.content['switch'].dpid
+        pipeline = self.get_enabled_table()
+        if pipeline.get('status') is not None:
+            self.install_miss_flows(pipeline, switch_id)
```
Error 424 (SwitchNotConnectedError) is occurring because `of_multi_table` tries to install miss flow entries before `of_core` listens to the `"kytos/of_core.handshake.completed"` event. Adding a `sleep(5)` before sending the miss flow entries solves the problem.

viniarck replied:
Good observation, I made a new comment regarding the handshake we can simplify things here, check out the other comments.

Alopalao replied:
Deleted

viniarck commented on main.py:
```diff
@@ -118,9 +313,56 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
 
+    def enable_basic_pipeline(self):
+        """Return to basic pipeline
+        After disabling a pipeline"""
```
Looks good, but can you build the content from `self.basic_pipeline`? Otherwise we'd be duplicating where the base pipeline is defined, years down the road, someone might forget to update here.

Alopalao replied:
Changed

viniarck commented on main.py:
```diff
@@ -118,9 +313,56 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
+    def handle_flow_mod_added(self, event):
+        """Handle recently added flows"""
+
+    @listen_to("kytos/of_core.handshake.completed")
```
In retrospect, based on what we specified in the blueprint, I don't think we actually need to be concerned with `"kytos/of_core.handshake.completed"`. Let me explain what I mean. table-miss flow entries, they'll only potentially exist when a new pipeline is defined, since the basic one won't have any, and you're covering it on `get_flows_to_be_installed()`.  We only would need to be concerned with `"kytos/of_core.handshake.completed"` if we needed to support this NApp without the consistency check on (because that way we wouldn't have guarantees that flow manager would always try to keep the flows installed, so when a handshake was completed then this NApp would need to push it again). So, with that said, I'd recommend to nuke subscribing to `"kytos/of_core.handshake.completed"`, and mention in the readme that this NApp needs `flow_manager.settings.ENABLE_CONSISTENCY_CHECK = True` (which is already on by default as you know), see what I mean? This also can make code maintenance easier since it's less code to maintain.

Alopalao replied:
Deleted listener and added requirement to readme

viniarck commented on main.py:
```diff
@@ -118,9 +313,56 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
+    def get_cookie(switch_dpid):
+        """Return the cookie integer given a dpid."""
+        dpid = int(switch_dpid.replace(":", ""), 16)
+        return (0x000FFFFFFFFFFFFF & dpid) | (COOKIE_PREFIX << 56)
```
Can you use here `0x00FFFFFFFFFFFFFF` instead? Preserving the last 7 bytes? As it is, it would also work (and we do have one NApp using this way), but since the cookie prefix only uses one byte, so we can also keep it a bit more consistent.

Alopalao replied:
Fixed

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+        flows = {
+          "flows": [
+            {
+              "cookie": 12465963768561532928,
```
The payload looks good, and `flow_manager` expects integer here, but can you also instead pass here a `hex` number but constructing it as `int(0x...`)? That way it's simpler for future readers to spot out the gate which mask and cookie is being used since reading in hex it's easier for cookie values and mask.

Alopalao replied:
Changed

viniarck commented on main.py:
```diff
@@ -97,12 +287,17 @@ def enable_pipeline(self, request: Request) -> JSONResponse:
             log.debug(f"enable_pipeline result {msg} 404")
             raise HTTPException(404, detail=msg)
-        msg = f"Pipeline {pipeline_id} enabled"
+        if pipeline.get("error") == "Conflict":
```
This looks very coherent, based on what we also discussed about returning a dict or optional, you can keep the logic on `update_status` to result in a conflict, that's OK too, but since this conflict here is "business" logic related for this NApp, it's becoming evident that we could benefit from exceptions. We also discussed in another review not to use http exceptions directly on controllers, so for this case, Aldo, you could add custom exceptions and introduce a new module `exceptions` where you could define a `PipelineBeingEnabled` or something along these lines or alternatively, you could also move that logic out of it (and then always make `get_pipeline` query first before trying to call `update_status(pipeline_id, "enabling")`, and I'll recommend that you move this part of the logic out of it and make another upfront query, otherwise if we end up in case where a pipeline hangs for ever being enabling (which typically won't happen, but it can happen depending on error handling or a NApp misbehave), then network engineers could also use the `/v1/pipeline/{pipeline_id}/disable` to try to rollback to the basic pipeline. Since it's desirable to also have this power when disabling a pipeline, then keep this conflict logic inside of `update_status` will be inconvenient, see what I mean?

Alopalao replied:
Separated into three methods. Now disabling a pipeline only will trigger the process of changing pipeline when the disabled pipeline was not `status: disabled` before.

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+                owner = flow["flow"]["owner"]
+                if owner not in set_up:
+                    continue
+                expected_table_id = set_up[owner][flow["flow"]["table_group"]]
```
Just a heads up that changes that I request on `flow_manager` will make the `flow["flow"]["table_group"]` to potentially trow a `KeyError`, can you also check if the value is truthy to be safe?

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+            # Pipeline enabling has finnished already
+            return
+        # Keys necessary to delete a flow, 'cookie_mask' is added later
+        keys = {'idle_timeout', 'hard_timeout', 'cookie', 'priority',
```
For flows deletion `flow_manager` sends a non-strict FlowMod, so you can further simplify here, and only send `cookie`, `cookie_mask` and `match`.

Alopalao replied:
Reduced.

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+                if expected_table_id != flow["flow"]["table_id"]:
+                    # Get key-value from flow to be sent to flow_manager
+                    delete = {key: flow["flow"].get(key) for key in keys}
+                    delete["cookie_mask"] = flow["flow"]["cookie"]
```
Since it's being deleted flow entry by flow entry, cookie mask should be all `0xf`'s, otherwise you can hit unexpected issues here.

Alopalao replied:
Fixed

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+        }
+        self.send_flows(flows, '', 'delete')
+
+    def send_flows(self, flows: Dict[str, list], switch: str, action: str):
```
You can replace this with the event-oriented interface and remove `requests` entirely from the dependencies too, check out my previous comments and if you agree with it, all things considered the event-oriented interface can simplify for this NApp which will be dealing with a large number of flows.

Alopalao replied:
Changed to events

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
     def execute(self):
         """Execute once when the napp is running."""
 
+    def get_enabled_table(self):
```
If you could also add the returned typing annotation, that'd be great, cosmetic detail since we don't have a linter for that yet, but it does enhance code readability out of the gate.

Alopalao replied:
Added annotations.

viniarck reviewed (changes requested):
@Alopalao excellent progress, it's shaping up very nicely. 

I asked some changes, most are implementation details though. 

It's fine to defer unit test to a subsequent PR, especially as you're collecting feedback, but once most of it it's working well and stabilizd let's make sure to add unit test covering at least 90%

viniarck commented on main.py:
```diff
@@ -31,11 +38,194 @@ def setup(self):
+        self.basic_pipeline = BASIC_PIPELINE
+        self.subscribed_napps = SUBSCRIBED_NAPPS
         self.pipeline_controller = self.get_pipeline_controller()
+        self._pipeline_lock = Lock()
```
You can also nuke the `self._pipeline_lock`, since we have DB guarantees when trying to enable a pipeline we won't need it, they way you're using `build_content` and passing the dict payload is thead-safe (it won't result in dict size changed during iteration)

Alopalao replied:
Deleted.

viniarck commented on main.py:
```diff
@@ -31,11 +38,232 @@ def setup(self):
+
+    def send_flows(self, flows: Dict, action: str, force: bool):
+        """Send flows to flow_manager through event"""
+        print("BATCH -> ", BATCH_SIZE)
```
Can you remove this temporary print statement?

Alopalao replied:
Deleted

viniarck commented on main.py:
```diff
@@ -31,11 +38,232 @@ def setup(self):
+            # Error when disabling pipeline
+            # Return to previous enabled pipeline
+            else:
+                pipeline = self.pipeline_controller.get_last_updated()
```
The idea of trying to return to a previous enabled pipeline might seem good, but I can see corner cases here. First, it's only relying on `updated_at`, which for now works, but if one day we also allow to update a pipeline description or minor updates, this will leave room for bugs if network operators update descriptions of all pipelines in an arbitrary order, so only relying on `updated_at` can lead to future surprises that we'll always need to be aware in code review. Second, what if a previous pipeline got disabled and it picks up an unwanted pipeline that network engineers weren't expecting to be activated anymore? 

It seems to me a risky move to try to rely on using `get_last_updated()` here, to simplify things and keep it deterministic can we always fall back to the default pipeline?

Alopalao replied:
No more roll back, more in this [comment](https://github.com/kytos-ng/of_multi_table/pull/13#discussion_r1199884843)

viniarck commented on main.py:
```diff
@@ -31,11 +38,232 @@ def setup(self):
+        if response.status_code // 100 != 2:
+            # Error when enabling pipeline
+            # Return to default pipeline
+            if pipeline.get("status") == "enabling":
```
Could you extract these statements that are dealing with error handling as new methods? This method `get_installed_flows` is having additional responsibilities that's not related to get the actual flows (so in practice it's a sort of a `get_installed_flows_or_rollback_pipeline`). The comments there help a bit, but it's also a sign that it should be a method to perform the error handling. Also, if you decorate it with `@retry` it'll become easier in the call site to try except `tenacity.RetryError` and call the new methods you'll extract accordingly, wdyt?

Alopalao replied:
There is no longer roll back. The pipelines will stay in its status and the user can retry with API request on the same pipeline. To avoid conflicts with other pipelines, I restricted other changes to other pipelines until there is not a `enabling` and `disabling` pipeline.

viniarck replied:
Good call. It minimizes edge cases and it's simpler to maintain. Let's close this thread.

viniarck commented on main.py:
```diff
@@ -31,11 +38,232 @@ def setup(self):
+    def get_installed_flows(self, pipeline: Dict) -> Optional[Dict]:
+        """Get flows from flow_manager"""
+        command = "v2/stored_flows?state=installed"
+        response = requests.get(f"{FLOW_MANAGER_URL}/{command}")
```
Can you replace `requests` with `httpx`? It turns out that `httpx` is now shipped on core, so we don't have to keep introducing `requests` as a NApp dependency anymore. [`httpx` has a similar API to `requests`](https://github.com/kytos-ng/kytos/blob/master/kytos/core/api_server.py#L224-L226) except that it also supports an async interface when needed too. It'd be great if you could also decorate this `get_installed_flows` with `@retry` from tenacity, retrying for `httpx.RequestError` up to 3 times. In the timeout request 20 secs should be fine, something similar that we've introduced on [`sdntrace_cp`](https://github.com/kytos-ng/sdntrace_cp/blob/master/utils.py#L13-L17), that's also calling this endpoint. If `flow_manager` is experiencing some extra load these retries can end up recovering from a transient error due to a potential heavy load.

Alopalao replied:
Replaced and added `@retry`, let me know if it is applied well.

viniarck replied:
Looks good, there's a last remaining part remaining though to wrap it up: https://github.com/kytos-ng/of_multi_table/pull/13#discussion_r1201132274

viniarck commented on main.py:
```diff
@@ -31,11 +38,232 @@ def setup(self):
+            delete_flows[switch] = [flow]
+        self.send_flows(delete_flows, 'delete', True)
+
+    def send_flows(self, flows: Dict, action: str, force: bool):
```
It's fantastic that you're levering the `force` option to rely on consistency check and simplify error handling in the call site. Since it's being used for both types of commands, can you set `force=True` by default? That way all `self.send_flows(install_flows)` don't have to keep passing `True`? Cosmetic detail, but that'd be great.

Alopalao replied:
Changed.

viniarck reviewed (commented):
@Alopalao, great work. It's almost there. The local tests you performed are great too. Since this PR is also the last major part I'll also wait for e2e tests to be passing before approving it, so if you could also implement the e2e part too and post the result here that'd be great.

viniarck commented:
@Alopalao, finally, it's not an explicit requirement of this PR, but I'd like to ask you to simulate changing a pipeline that has at least 500+ flows installed, on e2e tests you don't have to have a massive number of flows, but we should try to simulate at least once how it'll behave when dealing with a number of flows that we'll have in prod. You can also map a new task for this, whichever you prefer, it won't be a blocker for this PR, but we'll need to know the outcome to consider this NApp stable and add a stable badge in the README.md, signalling that it's ready for production usage.

viniarck commented:
@Alopalao, if you could also simulate if it'll handle correctly OFPT_ERROR that'd be great, you could try installing a pipeline with a flow miss entry that will result in ofpt_error, for instance trying to set `push_int` action on OvS (since this is only supported on NoviFlow, if you also have `noviflow` NApp running), it'll result in OFPT_ERROR 

```
{
    "flows": [
        {
            "priority": 100,
            "match": {
                "in_port": 1
            },
            "actions": [
                {
                    "action_type": "push_int"
                }
            ]
        }
    ]
}
```

In this case it'll be an unrecoverable error so it should also rollback to the default pipeline

Alopalao commented:
Updated first comment after commit [37aadeb77af67e77d1e02c4e9f44e5c87f79862e](https://github.com/kytos-ng/of_multi_table/pull/13/commits/37aadeb77af67e77d1e02c4e9f44e5c87f79862e)

Unit test are completed.
There is no longer error 424
Added new status, `disabling`:
- This value indicates that the pipeline is in process to be disabled but not finished yet. If recognized at start of Kytos, it will try again to disable it and apply the default pipeline.

viniarck commented on main.py:
```diff
@@ -31,11 +43,216 @@ def setup(self):
+        command = "v2/stored_flows?state=installed"
+        response = httpx.get(f"{FLOW_MANAGER_URL}/{command}")
+
+        if response.status_code // 100 != 2:
```
Could you replace this conditional with: `if response.is_server_error:`, and then `raise RequestError(message=f"{request.text} on "{command}")` with a message that carries some context? If you raise `RequestError` then you can benefit from the `retry` that will catch this exception type.

Alopalao replied:
Changed.

viniarck commented on main.py:
```diff
@@ -31,11 +43,216 @@ def setup(self):
+    def get_installed_flows(self) -> Optional[Dict]:
+        """Get flows from flow_manager"""
+        command = "v2/stored_flows?state=installed"
+        response = httpx.get(f"{FLOW_MANAGER_URL}/{command}")
```
```suggestion
        response = httpx.get(f"{FLOW_MANAGER_URL}/{command}", timeout=20)
```

By default, `httpx` uses 5 secs as timeout, let's parametrize 20 secs or so here, just so it might also succeed in the first try when the traffic might be spiky.

Alopalao replied:
Added

viniarck commented on main.py:
```diff
@@ -31,11 +43,216 @@ def setup(self):
+        try:
+            flows_by_swich = self.get_installed_flows()
+        except tenacity.RetryError as err:
+            raise HTTPException(424, "It couldn't get stored_flows") from err
```
This error handling here can still blows up. This is being called from a `@listen_to("kytos/(mef_eline|coloring|of_lldp).enable_table")` thread pool handler, so if you raise `HTTPException` since it's not in a request handler it won't catch it. Since we're handling the pipeline completely out of the request cycle, then at this point here, it should probably:

- `log.error` since it won't recover from this until a new request is made, and if this happens it means the back-end server is experiencing issues and it would need attention anyway, so `log.error` is appropriate here. 
- It might be worth updating `status` as either `disabling_error` or `enabling_error` (maybe these two would be the easiest ones to play well with how you structured - or something along these lines - where the error would be visible to the user - but then you need to take into account that would be a valid state when trying to enable or disable the pipeline again), just so from the API they can know that the back-end has reached maximum retries and will no longer try to recover from this.

viniarck replied:
Let me know if you see any other better and smoother way to handle and expose this

Alopalao replied:
Added new status and `log.error`

viniarck commented on controllers/__init__.py:
```diff
@@ -72,20 +81,57 @@ def get_pipeline(self, id: str) -> Optional[Dict]:
+        )
+        return pipeline
+    
+    def get_last_updated(self) -> Dict:
```
Since this method is no longer used, can you delete it?

Alopalao replied:
Deleted

viniarck commented on setup.cfg:
```diff
@@ -3,7 +3,7 @@ exclude = .eggs,ENV,build,docs/conf.py,venv
 [yala]
 linters=pylint,pycodestyle,isort
-pylint args = --disable=too-few-public-methods,too-many-instance-attributes,no-name-in-module,unnecessary-pass --ignored-modules=napps.amlight.coloring
+pylint args = --disable=too-few-public-methods,too-many-instance-attributes,no-name-in-module,unnecessary-pass,missing-timeout --ignored-modules=napps.amlight.coloring
```
Can you revert the `missing-timeout` ignored lint error from `requests` since now you're using `httpx` and parametrizing the timeout?

Alopalao replied:
Removed

viniarck commented on controllers/__init__.py:
```diff
@@ -49,7 +49,16 @@ def insert_pipeline(self, pipeline: Dict) -> InsertOneResult:
+    def get_active_pipeline(self) -> Dict:
+        """Get a pipeline that is enabling, disabling or enabled.
+        There can only be only be one of them at a time"""
+        return self.db.pipelines.find_one({"$or":[
```
Could you rewrite this query using the `"$in": ["enabling", "enabled", "disabling"]` , operator instead? Usually the `$or` operator fits well when the query as disjoints  like a `UNION` on sql, `"$in"` would be more idiomatic here. Also, on MongoDB forums an employee [has pointed out that usually `$in` tends to perform better for most use cases](https://www.mongodb.com/community/forums/t/performance-of-or-vs-in/175513/5), for our case this collection wouldn't be that large, but still `$in` would be more idiomatic here

Alopalao replied:
Rewritten, now it will look for something different than `disabled`

viniarck commented on main.py:
```diff
@@ -2,17 +2,29 @@
 
+import httpx
+import tenacity
+from httpx._exceptions import RequestError
```
```suggestion
from httpx import RequestError
```

You can import it directly from `httpx` instead of a "private" module

Alopalao replied:
Fixed

viniarck commented on controllers/__init__.py:
```diff
@@ -72,20 +81,57 @@ def get_pipeline(self, id: str) -> Optional[Dict]:
-    def update_status(self, id: str, status: str) -> Optional[Dict]:
-        """Update the status of a pipeline"""
+
+    def enabling_pipeline(self, id: str) -> Optional[Dict]:
```
`"enabling"` and `"disabling"` aren't documented in the openapi.yml yet, also since the number of status is increasing (which is fine), it's probably worth introducing an Enumerable when setting the values, just so we also have a clear way to know all the possible values easily that we're currently supporting something like:

```
class PipelineStatus(Enum):
    """Enum for pipeline status."""

    ENABLED = "enabled"
    ...
```

wdyt?

Alopalao replied:
Added new status to openapi.yml and added Enum class (missing error status)

viniarck commented on main.py:
```diff
@@ -31,11 +43,216 @@ def setup(self):
+                    }
+                    if flow["flow"].get('match'):
+                        delete['match'] = flow["flow"].get('match')
+                    delete_flows[switch].append(delete.copy())
```
The shallow copy isn't expensive here, but do we really need this copy here? It's looking redundant and just extra to me, since `delete: dict` is already a unique ref  per iteration and won't be mutated.

Alopalao replied:
Removed.

viniarck commented on main.py:
```diff
@@ -31,11 +43,216 @@ def setup(self):
+                    delete_flows[switch].append(delete.copy())
+                    # Change table_id before being added
+                    flow["flow"].update({"table_id": expected_table_id})
+                    install_flows[switch].append(flow["flow"].copy())
```
Do we need this shallow copy here as well? Looks redundant to me since at this point forward `flow["flow"]` is no longer mutated, let me know if you've caught any issue or has a reason that I missed.

Alopalao replied:
Removed, it was a leftover of a previous implementation.

viniarck commented on main.py:
```diff
@@ -111,16 +342,47 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
         log.debug(f"disable_pipeline result {msg} 200")
         return JSONResponse(msg)
 
+    @listen_to("kytos/flow_manager.flow.added")
```
Looks like subscribing to `"kytos/flow_manager.flow.added"` since this NApp is also relying on consistency check to install the flows in the worst case and isn't keeping track of each and individual flow that's been confirmed, can we removed this in this iteration?

Alopalao replied:
Deleted

viniarck commented on main.py:
```diff
@@ -111,16 +342,47 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
+    @listen_to("kytos/flow_manager.flow.error")
+    def on_flow_mod_error(self, event):
+        """Handle flow mod errors"""
+        self.handle_flow_mod_error(event)
```
`"kytos/flow_manager.flow.error"` isn't still being handled, if a user install a table-miss flow poorly formatted or with unsupported action the pipeline might not end up fully functional. At this point, since we're no longer rolling back should it should update the status of the pipeline as `enabling_error`? Let me know if you had other things in mind, it's important to simulate an actual OFPT_ERROR and see how it'll behave, one day a network engineer will exercise this path here, either accidentally or not :)

Alopalao replied:
Partially handled. When the error is produced by a miss flow, the pipeline status will change from `enabled` to `enabling_error`. Missing case when the table id is not accepted by the switch.

viniarck replied:
Overall, looks good. You'll need to try to simulate at least once an `OFPT_ERROR` though to see this working at least once in practice, let me know once you complete this part.

Alopalao replied:
Checked with:
```
{
    "flows": [
        {
            "cookie": 12465963768561532929,
            "priority": 100,
            "match": {
                "in_port": 1
            },
            "actions": [
                {
                    "action_type": "push_int"
                }
            ]
        }
    ]
}
```

viniarck replied:
Cool. Let's close this thread.

viniarck reviewed (commented):
@Alopalao thanks for the updates, I've just opened new follow up threads. Good call simplifying and not supporting the auto-rollback, it's way to maintain now.

viniarck reviewed (commented):
@Alopalao thanks for the updates, I've just opened new follow up threads. Good call simplifying and not supporting the auto-rollback, it's way easier to maintain now.

viniarck reviewed (commented):
@Alopalao thanks for the updates, I've just opened new follow up threads. Good call simplifying and not supporting the auto-rollback, it's way easier to maintain now.

Alopalao commented:
Commit [3a56caf0c94986bd6fc50fa5552c5eec27ba6798](https://github.com/kytos-ng/of_multi_table/pull/13/commits/3a56caf0c94986bd6fc50fa5552c5eec27ba6798)

Tox failing because `error_msg` is not recognized as attribute from `kytos.core.rest_api`. So I haven't committed the tests for now.

With the increasing number of pipeline status. I tested so enabling pipeline is more dynamic. Now it is no longer necessary to have everything `disabled` (default pipeline) before trying to enable a pipeline. This comes with the tiny cost of always trying to delete miss flows before installing new miss flows when necessary.

Missing catching error when a switch does not allow a certain table_id.

viniarck commented on openapi.yml:
```diff
@@ -209,4 +206,6 @@ components:
             - disabled
-            - enabled
\ No newline at end of file
+            - enabled
```
`enabling_error` and `disabling_error` are still missing here, can you augment this?

Alopalao replied:
Added.

viniarck commented on main.py:
```diff
@@ -111,26 +338,43 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
+        self.handle_flow_mod_error(event)
+
+    def handle_flow_mod_error(self, event):
+        """Handle flow mod errors"""
```
As documented [on `flow_manager`](https://github.com/kytos-ng/flow_manager#kytosflow_managerflowerror), if `event.content.get("error_exception")` is set then it's not an `OFPT_ERROR` so you can return early in that case.

Alopalao replied:
Added early return

viniarck commented on main.py:
```diff
@@ -111,26 +338,43 @@ def disable_pipeline(self, request: Request) -> JSONResponse:
+
+    def handle_flow_mod_error(self, event):
+        """Handle flow mod errors"""
+        flow = event.content["flow"]
```
`flow` here is an instance of `Flow04`, so `flow["owner"]` below will crash when an `OFPT_ERROR` actually happens, you need to access `flow.cookie` and check if belongs to `of_multi_table`. Also, consequently, you probably want to use `flow.as_dict()` in the `log.error` since `Flow04` doesn't have a custom `__repr__`, could you adapt these parts accordingly?

Alopalao replied:
Changed

viniarck commented on setup.cfg:
```diff
@@ -3,7 +3,7 @@ exclude = .eggs,ENV,build,docs/conf.py,venv
 [yala]
 linters=pylint,pycodestyle,isort
-pylint args = --disable=too-few-public-methods,too-many-instance-attributes,no-name-in-module,unnecessary-pass --ignored-modules=napps.amlight.coloring
+pylint args = --disable=too-few-public-methods,too-many-instance-attributes,no-name-in-module,unnecessary-pass,missing-timeout --ignored-modules=napps.amlight.coloring
```
missing-timeout is still here, can you could check?

Alopalao replied:
Now it is delete.

viniarck commented on main.py:
```diff
@@ -31,11 +43,221 @@ def setup(self):
+            flows_by_swich = self.get_installed_flows()
+        except tenacity.RetryError:
+            status = pipeline.get("status")
+            status = "enabling_error" if status else "disabling_error"
```
Let's suppose a pipeline is currently in `disabling_error`, and then a user tries to make a request to  `POST /v1/pipeline/{pipeline_id}/disable`, which will set the pipeline status again as `disabling` (which is as expected), since it's trying to potentially fix it assuming it was a recoverable error like when retries fails here for instance (extremely rare but it can happen), in this case the prior state would be `disabling`, so `if status` would be true and it would result in `enabling_error`, which wouldn't be coherent, right? Did I follow this correctly? Can you double check?

Alopalao replied:
`pipeline` is replace with `self.default_pipeline` before when a status has the str `disabl` in its value, [line](https://github.com/kytos-ng/of_multi_table/blob/3a56caf0c94986bd6fc50fa5552c5eec27ba6798/main.py#L150-L151). `status` can only get `enabling` and `enabling_error` from the dictionary.

viniarck replied:
Oh, I see, Aldo. Right. Let's close this thread then.

viniarck reviewed (changes requested):
@Alopalao, great progress. I closed most of the prior threads, and followed up. 

Regarding:

> Tox failing because error_msg is not recognized as attribute from kytos.core.rest_api. So I haven't committed the tests for now.

If you could fix that part too it'd be great since you're PR has landed on `kytos`. Other unit tests are failing too, which makes it unclear if you're still trying to make other changes before a final review, ping me, when you need a final review:

```
FAILED tests/unit/test_controllers.py::TestController::test_get_active_pipeline - KeyError: '$or'
FAILED tests/unit/test_main.py::TestMain::test_get_flows_to_be_installed - AssertionError: assert 2 == 1
FAILED tests/unit/test_main.py::TestMain::test_delete_pipeline - assert 409 == 200
FAILED tests/unit/test_main.py::TestMain::test_enable_pipeline_conflict - assert 200 == 409
FAILED tests/unit/test_main.py::TestMain::test_disable_pipeline_not_found - assert 200 == 404
FAILED tests/unit/test_main.py::TestMain::test_disable_pipeline_conflict - KeyError: 'id'
FAILED tests/unit/test_main.py::TestMain::test_error_msg - AttributeError: 'Main' object has no attribute 'error_msg'
```

Alopalao commented:
Commit [5e8c4857c3bed9b6a6249c7e223c455f141b01f3](https://github.com/kytos-ng/of_multi_table/pull/13/commits/5e8c4857c3bed9b6a6249c7e223c455f141b01f3)

Tests are working normally now.
Updated first comment with latest changes.

viniarck commented:
@Alopalao, looks good. 

I was also exploring this locally, it seems solid, but I encountered a case when I disabled the pipeline, and although it deleted the flows from the DB correctly, in the switch it temporarily stayed there longer, but the consistency check ended up cleaning them up. It might be something related to a similar match but in different tables, if you could look into it. Here's the case:

- I created this pipeline, to simulate moving an EVC (epl) from table 0 to table 1:

```
{
    "multi_table": [
        {
            "table_id": 0,
            "description": "table_zero",
            "table_miss_flow": {
                "instructions": [
                    {
                        "instruction_type": "goto_table",
                        "table_id": 1
                    }
                ],
                "priority": 0,
                "match": {}
            },
            "napps_table_groups": {
                "coloring": [
                    "base",
                    "base"
                ],
                "of_lldp": [
                    "base",
                    "base"
                ]
            }
        },
        {
            "table_id": 1,
            "description": "table_one",
            "napps_table_groups": {
                "mef_eline": [
                    "epl",
                    "evpl"
                ]
            }
        }
    ]
}
```

- Here's the flows in the switch before enabling the pipeline:

```
❯ sudo ovs-ofctl -O OpenFlow13 dump-flows s1 
 cookie=0xac00000000000001, duration=215.259s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=50000,dl_src=ee:ee:ee:ee:ee:03 actions=CONTROLLER:65535
 cookie=0xac00000000000001, duration=215.255s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=50000,dl_src=ee:ee:ee:ee:ee:02 actions=CONTROLLER:65535
 cookie=0xaafc9ed85266ef41, duration=209.559s, table=0, n_packets=8, n_bytes=592, send_flow_rem priority=20000,in_port="s1-eth4",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=209.459s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=20000,in_port="s1-eth3",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=209.560s, table=0, n_packets=8, n_bytes=560, send_flow_rem priority=10000,in_port="s1-eth1" actions=push_vlan:0x88a8,set_field:4097->vlan_vid,output:
"s1-eth4"
 cookie=0xab00000000000001, duration=216.696s, table=0, n_packets=288, n_bytes=12096, send_flow_rem priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
```

- Here's the flows in the switch after enabling the pipeline (looks good):

```
❯ sudo ovs-ofctl -O OpenFlow13 dump-flows s1 
 cookie=0xac00000000000001, duration=286.236s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=50000,dl_src=ee:ee:ee:ee:ee:03 actions=CONTROLLER:65535
 cookie=0xac00000000000001, duration=286.232s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=50000,dl_src=ee:ee:ee:ee:ee:02 actions=CONTROLLER:65535
 cookie=0xab00000000000001, duration=287.673s, table=0, n_packets=384, n_bytes=16128, send_flow_rem priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
 cookie=0xad00000000000001, duration=57.581s, table=0, n_packets=34, n_bytes=3320, send_flow_rem priority=0 actions=goto_table:1
 cookie=0xaafc9ed85266ef41, duration=58.591s, table=1, n_packets=13, n_bytes=1186, send_flow_rem priority=20000,in_port="s1-eth4",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=58.590s, table=1, n_packets=0, n_bytes=0, send_flow_rem priority=20000,in_port="s1-eth3",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=58.589s, table=1, n_packets=13, n_bytes=1134, send_flow_rem priority=10000,in_port="s1-eth1" actions=push_vlan:0x88a8,set_field:4097->vlan_vid,output
:"s1-eth4"
```

- But, once I disabled the pipeline, to go back to the default one, it still kept some flows in the switch on table=1, although in the DB they were deleted correctly (it might be a matter of also sending the table_id in the flows too):

```
kytos $> 2023-05-30 20:07:35,793 - INFO [uvicorn.access] [httptools_impl.py:506:send] (MainThread) 127.0.0.1:35962 - "POST /api/kytos/of_multi_table/v1/pipeline/2d6130ac2fc04a51ae733fe8
9a61223e/disable HTTP/1.1" 200
2023-05-30 20:07:35,814 - INFO [uvicorn.access] [httptools_impl.py:506:send] (MainThread) 127.0.0.1:35976 - "GET /api/kytos/flow_manager/v2/stored_flows?state=installed HTTP/1.1" 200
2023-05-30 20:07:35,816 - INFO [kytos.napps.kytos/flow_manager] [main.py:562:handle_flows_install_delete] (thread_pool_app_1) Send FlowMod from KytosEvent dpid: 00:00:00:00:00:00:00:01,
 command: delete, force: True, flows_dict: {'flows': [{'cookie': 12320897332511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 4, 'dl_vlan': 1}}, {'cookie': 1232089733
2511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 3, 'dl_vlan': 1}}, {'cookie': 12320897332511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 1}}]}
2023-05-30 20:07:35,818 - INFO [kytos.napps.kytos/flow_manager] [main.py:562:handle_flows_install_delete] (thread_pool_app_6) Send FlowMod from KytosEvent dpid: 00:00:00:00:00:00:00:03,
 command: delete, force: True, flows_dict: {'flows': [{'cookie': 12320897332511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 3, 'dl_vlan': 1}}, {'cookie': 1232089733
2511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 2, 'dl_vlan': 1}}, {'cookie': 12320897332511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 1}}]}
2023-05-30 20:07:35,821 - INFO [kytos.napps.kytos/flow_manager] [main.py:562:handle_flows_install_delete] (thread_pool_app_5) Send FlowMod from KytosEvent dpid: 00:00:00:00:00:00:00:02,
 command: delete, force: True, flows_dict: {'flows': [{'cookie': 12320897332511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 2, 'dl_vlan': 1}}, {'cookie': 1232089733
2511436609, 'cookie_mask': 18446744073709551615, 'match': {'in_port': 3, 'dl_vlan': 1}}]}
```


```
❯ sudo ovs-ofctl -O OpenFlow13 dump-flows s1 
 cookie=0xac00000000000001, duration=363.446s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=50000,dl_src=ee:ee:ee:ee:ee:03 actions=CONTROLLER:65535
 cookie=0xac00000000000001, duration=363.442s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=50000,dl_src=ee:ee:ee:ee:ee:02 actions=CONTROLLER:65535
 cookie=0xaafc9ed85266ef41, duration=39.547s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=20000,in_port="s1-eth4",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=39.544s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=20000,in_port="s1-eth3",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=39.541s, table=0, n_packets=0, n_bytes=0, send_flow_rem priority=10000,in_port="s1-eth1" actions=push_vlan:0x88a8,set_field:4097->vlan_vid,output:"s1
-eth4"
 cookie=0xab00000000000001, duration=364.883s, table=0, n_packets=484, n_bytes=20328, send_flow_rem priority=1000,dl_vlan=3799,dl_type=0x88cc actions=CONTROLLER:65535
 cookie=0xaafc9ed85266ef41, duration=135.801s, table=1, n_packets=39, n_bytes=3782, send_flow_rem priority=20000,in_port="s1-eth4",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=135.800s, table=1, n_packets=0, n_bytes=0, send_flow_rem priority=20000,in_port="s1-eth3",dl_vlan=1 actions=pop_vlan,output:"s1-eth1"
 cookie=0xaafc9ed85266ef41, duration=135.799s, table=1, n_packets=39, n_bytes=3626, send_flow_rem priority=10000,in_port="s1-eth1" actions=push_vlan:0x88a8,set_field:4097->vlan_vid,outpu
t:"s1-eth4"


```

Only after some delta t, the consistency check cleaned them up, so it still worked as expected from a functionality point of view, but these ones here should've been deleted by `of_multi_table` when the pipeline was being disabled and not on consistency check, the cases where we expect to see consistency check are only for indeed anomalies like when a flow is deleted in the switch accidentally or if a connectivity issue happens:

```
kytos $> 2023-05-30 20:08:55,300 - INFO [uvicorn.access] [httptools_impl.py:506:send] (MainThread) 127.0.0.1:58030 - "GET /api/kytos/of_multi_table/v1/pipeline HTTP/1.1" 200            
2023-05-30 20:10:42,916 - INFO [kytos.napps.kytos/flow_manager] [main.py:417:check_alien_flows] (thread_pool_sb_9) Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
2023-05-30 20:10:42,918 - INFO [kytos.napps.kytos/flow_manager] [main.py:422:check_alien_flows] (thread_pool_sb_9) Flow forwarded to switch 00:00:00:00:00:00:00:01 to be deleted. Flow: 
{'flows': [{'switch': '00:00:00:00:00:00:00:01', 'table_id': 1, 'match': {'in_port': 4, 'dl_vlan': 1}, 'priority': 20000, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 123208973325114
36609, 'id': '58dc1a024a4ddad71f4791ced3f06a83', 'stats': {'byte_count': 3782, 'duration_sec': 282, 'duration_nsec': 787000000, 'packet_count': 39}, 'cookie_mask': 0, 'instructions': [{
'instruction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'port': 1, 'action_type': 'output'}]}]}]}
2023-05-30 20:10:42,918 - INFO [kytos.napps.kytos/flow_manager] [main.py:417:check_alien_flows] (thread_pool_sb_9) Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
2023-05-30 20:10:42,919 - INFO [kytos.napps.kytos/flow_manager] [main.py:422:check_alien_flows] (thread_pool_sb_9) Flow forwarded to switch 00:00:00:00:00:00:00:01 to be deleted. Flow: 
{'flows': [{'switch': '00:00:00:00:00:00:00:01', 'table_id': 1, 'match': {'in_port': 3, 'dl_vlan': 1}, 'priority': 20000, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 123208973325114
36609, 'id': 'da677bc9799e8ea4e2ac1c76d69aa1e3', 'stats': {'byte_count': 0, 'duration_sec': 282, 'duration_nsec': 786000000, 'packet_count': 0}, 'cookie_mask': 0, 'instructions': [{'ins
truction_type': 'apply_actions', 'actions': [{'action_type': 'pop_vlan'}, {'port': 1, 'action_type': 'output'}]}]}]}
2023-05-30 20:10:42,920 - INFO [kytos.napps.kytos/flow_manager] [main.py:417:check_alien_flows] (thread_pool_sb_9) Consistency check: alien flow on switch 00:00:00:00:00:00:00:01
2023-05-30 20:10:42,921 - INFO [kytos.napps.kytos/flow_manager] [main.py:422:check_alien_flows] (thread_pool_sb_9) Flow forwarded to switch 00:00:00:00:00:00:00:01 to be deleted. Flow: 
{'flows': [{'switch': '00:00:00:00:00:00:00:01', 'table_id': 1, 'match': {'in_port': 1}, 'priority': 10000, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 12320897332511436609, 'id': '
5002e176800ef83a2124c7394d638213', 'stats': {'byte_count': 3626, 'duration_sec': 282, 'duration_nsec': 785000000, 'packet_count': 39}, 'cookie_mask': 0, 'instructions': [{'instruction_t
ype': 'apply_actions', 'actions': [{'action_type': 'push_vlan', 'tag_type': 's'}, {'vlan_id': 1, 'action_type': 'set_vlan'}, {'port': 4, 'action_type': 'output'}]}]}]}
2023-05-30 20:10:48,929 - INFO [kytos.napps.kytos/flow_manager] [main.py:417:check_alien_flows] (thread_pool_sb_7) Consistency check: alien flow on switch 00:00:00:00:00:00:00:02
2023-05-30 20:10:48,931 - INFO [kytos.napps.kytos/flow_manager] [main.py:422:check_alien_flows] (thread_pool_sb_7) Flow forwarded to switch 00:00:00:00:00:00:00:02 to be deleted. Flow: 
{'flows': [{'switch': '00:00:00:00:00:00:00:02', 'table_id': 1, 'match': {'in_port': 2, 'dl_vlan': 1}, 'priority': 20000, 'idle_timeout': 0, 'hard_timeout': 0, 'cookie': 123208973325114
36609, 'id': 'de10f0588c48821525949000c8db1e90', 'stats': {'byte_count': 0, 'duration_sec': 288, 'duration_nsec': 792000000, 'packet_count': 0}, 'cookie_mask': 0, 'instructions': [{'ins
truction_type': 'apply_actions', 'actions': [{'vlan_id': 1, 'action_type': 'set_vlan'}, {'port': 3, 'action_type': 'output'}]}]}]}
...
```

viniarck commented:
@Alopalao, I'll approve once the last comment is addressed and when the e2e tests that were requested on the other e2e PR that you're working on are passing. Nicely done, just final touches now :)

Also, let me know when you'll cover the validation to ensure the pipeline tables are always ascending and loop free. In addition, looks like it's possible to use the same table group in multiple tables which could lead to issues. 

Finally, I'd like to also know if you stress tested this with at least 500+ flows to see the batch in actions and smoothly transitioning, which will pretty much be the case when this will be used in production, this second part doesn't need to be done here, it's up to you if you want to address or cover in a new task, let me know what you prefer.

Alopalao commented:
Commit [7138a91d8d733ba8f01447b952421e40ae8339f7](https://github.com/kytos-ng/of_multi_table/pull/13/commits/7138a91d8d733ba8f01447b952421e40ae8339f7)

- Added validators to pipeline document so it checks that there is not duplicated table groups from a napp and the table id from instructions always goes up ensuring no loops.
- Fixed the necessity for consistency check when disabling pipeline
- Testing activation of pipeline with ~500 flows

Commit [3b685f7de570630c1e5461017cf11d73a8d41c79](https://github.com/kytos-ng/of_multi_table/pull/13/commits/3b685f7de570630c1e5461017cf11d73a8d41c79)
- Added validation to find `table_id` duplication.

viniarck reviewed (approved):
LGTM. Nicely done @Alopalao 

Let's ship it. Since e2e tests PR is also progressing well, and this NApp doesn't have e2e yet, it's OK to land this first without waiting for it.

![img](https://media.giphy.com/media/Hw8vYF4DNRCKY/giphy.gif)
